### ISSUE TITLE
Writing a VCF with no variants omits header (regression introduced in 0.10.0)

### ISSUE THREAD

**Issue by kghosesbg** [pysam-developers/pysam#450]
When writing out a VCF, if there are no variant rows written then the header is not written either. This is unexpected behavior that was introduced after 0.9.0 (which operates as expected) and is present in 0.10.0 and 0.11.0.

(In the example below `input.vcf.gz` is any sample VCF you have lying around)
```
import pysam

print(pysam.__version__)
vcf_in = pysam.VariantFile('input.vcf.gz', 'r')
vcf_out = pysam.VariantFile('test.vcf', 'w', header=vcf_in.header)
```
Running this with 0.9.0 produces the expected result

```
$ python pysam-vcf-bug.py 
0.9.0
$ cat test.vcf 
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##fileDate=20160824
##contig=<ID=1,length=249250621>
##contig=<ID=2,length=243199373>
##contig=<ID=3,length=198022430>
##contig=<ID=4,length=191154276>
##contig=<ID=5,length=180915260>
##contig=<ID=6,length=171115067>
##contig=<ID=7,length=159138663>
##contig=<ID=8,length=146364022>
##contig=<ID=9,length=141213431>
##contig=<ID=10,length=135534747>
##contig=<ID=11,length=135006516>
##contig=<ID=12,length=133851895>
##contig=<ID=13,length=115169878>
##contig=<ID=14,length=107349540>
##contig=<ID=15,length=102531392>
##contig=<ID=16,length=90354753>
##contig=<ID=17,length=81195210>
##contig=<ID=18,length=78077248>
##contig=<ID=19,length=59128983>
##contig=<ID=20,length=63025520>
##contig=<ID=21,length=48129895>
##contig=<ID=22,length=51304566>
##contig=<ID=X,length=155270560>
##contig=<ID=Y,length=59373566>
##contig=<ID=MT,length=16569>
##FORMAT=<ID=GT,Number=1,Type=String,Description="Consensus Genotype across all datasets with called genotype">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	ref
```

But for later versions of pysam, we get an empty file
```
$ python pysam-vcf-bug.py 
0.10.0 (or 0.11.0)
$ cat test.vcf 
$ 
```

Adding a `vcf_out.close()` statement results in a segfault on 0.10.0 but on 0.11.0 will write out the header.

Thanks!
-Kaushik

**Comment by bioinformed** [comment]
The segfault was due to a bug present in `0.10.0`, but fixed in `0.11.0`.  It is necessary to close `vcf_out`, since the header is written lazily.  It is this way or else it wouldn't be possible to modify the header after creating `vcf_out` and before writing the first records.

**Comment by bioinformed** [comment]
Please do verify that you see the segfault with pysam 0.11.0.  Thanks!

**Comment by kghosesbg** [comment]
@bioinformed Aak, you are correct, sir, the segfault does not happen with 0.11.0 and explicitly closing the file will write out the header on 0.11.0

I would request, though, that the `.close()` command be executed implicitly with object destruction. It's more Pythonic, than to have to call `.close()` explicitly.

Thanks!

**Comment by bioinformed** [comment]
Hmm.  You're right.  The `__dealloc__` method seems to have gotten lost in a recent refactor.  Thanks for the report -- I'll fix that ASAP.

**Comment by bioinformed** [comment]
Fixed in PR #456.

**Comment by bioinformed** [comment]
Fix release in v0.11.2.1
